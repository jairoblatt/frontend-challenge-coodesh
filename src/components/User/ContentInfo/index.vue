<template>
  <div class="content__info">
    <span> <Icon :icon="icon" size="19" /> </span>
    <small>
      <slot />
    </small>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  props: {
    icon: String as PropType<string>,
  },
});
</script>

<style lang="postcss" scoped>
.content__info {
  @apply flex items-center justify-start md:justify-center border-solid border rounded-md border-gray-100 my-1 md:my-3 mx-2 md:w-auto w-full;
}

.content__info span {
  @apply font-bold py-1 px-3 bg-primary-base text-white;
}

.content__info small {
  @apply font-semibold px-4 text-base;
}
</style>
