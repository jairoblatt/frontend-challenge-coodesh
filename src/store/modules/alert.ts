import { GetterTree, MutationTree, ActionTree } from "vuex";
import { IAlert } from "@/types/IAlert";

export const state = () => ({
  alerts: [] as IAlert[],
});

export type AlertState = ReturnType<typeof state>;

export const getters: GetterTree<AlertState, AlertState> = {
  get: (state) => state.alerts[0] || null,
};

export const mutations: MutationTree<AlertState> = {
  SET_ALERT(state, payload) {
    state.alerts.push(payload);
  },

  UPDATE_ALERT(state, payload) {
    state.alerts = payload;
  },
};

export const actions: ActionTree<AlertState, AlertState> = {
  set({ commit }, payload: IAlert) {
    const { timeOut } = payload;
    commit("SET_ALERT", {
      ...payload,
      timeOut: timeOut === false ? timeOut : Number(timeOut) > 1 ? timeOut : 10000,
      id: Math.random()
        .toString(36)
        .substr(2),
    });
  },

  remove({ commit, state }, alertId) {
    commit(
      "UPDATE_ALERT",
      state.alerts.filter(({ id }) => id !== alertId)
    );
  },
};
