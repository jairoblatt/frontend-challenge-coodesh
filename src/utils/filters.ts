export const genders = [
  {
    value: "",
    name: "All",
  },
  {
    value: "female",
    name: "Female",
  },
  {
    value: "male",
    name: "Male",
  },
];

export const nationalities = [
  {
    value: "",
    name: "All",
  },
  {
    value: "au",
    name: "Australian",
  },
  {
    value: "br",
    name: "Brazilian",
  },
  {
    value: "ca",
    name: "Canadian",
  },
  {
    value: "ch",
    name: "Switzerland",
  },
  {
    value: "de",
    name: "German",
  },
  {
    value: "dk",
    name: "Danish",
  },
  {
    value: "es",
    name: "Spanish",
  },
  {
    value: "fi",
    name: "Finnish",
  },
  {
    value: "fr",
    name: "French",
  },
  {
    value: "gb",
    name: "United Kingdom",
  },
  {
    value: "ie",
    name: "Irish",
  },
  {
    value: "ir",
    name: "Iraqi",
  },
  {
    value: "no",
    name: "Norwegian",
  },
  {
    value: "nl",
    name: "Dutch",
  },
  {
    value: "nz",
    name: "New Zealand",
  },
  {
    value: "tr",
    name: "Turkey",
  },
  {
    value: "us",
    name: "American",
  },
];
