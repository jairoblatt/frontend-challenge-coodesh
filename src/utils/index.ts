export * from "./sleep";
export * from "./filters";
