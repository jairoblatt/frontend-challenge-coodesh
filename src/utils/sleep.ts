export function sleep(milliseconds?: number) {
  return new Promise<void>((resolve) => {
    setTimeout((): void => resolve(), milliseconds || 1);
  });
}
