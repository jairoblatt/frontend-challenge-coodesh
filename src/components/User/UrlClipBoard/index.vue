<template>
  <div :class="classes">
    <input :value="locationHref" type="text" readonly @focus="copyToClipBoard" />
    <Btn data-testid="clipboard" @click="copyToClipBoard">
      <Icon :icon="iconClipBoard" />
    </Btn>
  </div>
</template>
<script lang="ts">
import Vue from "vue";
import { sleep } from "@/utils";

export default Vue.extend({
  data: () => ({
    copied: false,
  }),

  computed: {
    classes(): { [key: string]: boolean } {
      return {
        user__url: true,
        "user__url--copied": this.copied,
      };
    },

    locationHref(): string {
      return window.location.href;
    },

    iconClipBoard(): string {
      return this.copied ? "mdi-check" : "mdi-content-copy";
    },
  },

  methods: {
    async copyToClipBoard() {
      if (!this.copied) {
        await navigator.clipboard.writeText(this.locationHref);
        this.copied = true;

        await sleep(5000);
        this.copied = false;
      }
    },
  },
});
</script>
<style lang="postcss" scoped>
.user__url {
  @apply border border-light-onSurfacePrimary w-11/12 md:w-2/3 overflow-hidden border-opacity-40 border-solid rounded-sm mt-2 flex justify-between items-center;
}

.user__url input {
  @apply outline-none font-semibold px-2 w-full;
}
.user__url >>> button {
  @apply text-light-onSurfacePrimary border-l rounded-none border-light-onSurfacePrimary border-opacity-40 border-solid hover:bg-light-onSurfacePrimary hover:text-white;
}

.user__url--copied {
  @apply border-primary-base;
}

.user__url--copied >>> button {
  @apply border-primary-base hover:bg-primary-base bg-primary-base text-white;
}
</style>
