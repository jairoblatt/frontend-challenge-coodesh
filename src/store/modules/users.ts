import { GetterTree, MutationTree, ActionTree } from "vuex";
import HTTPClient from "@/plugins/axios";
import { Nullable, Users, UserApiInfo } from "@/types/IUser";
import { sleep } from "@/utils";

export const state = () => ({
  users: null as Nullable<Users[]>,
  fetchInfo: null as Nullable<UserApiInfo>,
  fetchPending: false,
});

export type UserState = ReturnType<typeof state>;

export const getters: GetterTree<UserState, UserState> = {
  get: (state) => state.users,
  fetchInfo: (state) => state.fetchInfo,
  fetchPending: (state) => state.fetchPending,
};

export const mutations: MutationTree<UserState> = {
  SET_USERS(state, payload) {
    state.users = payload;
  },

  FETCH_PENDING(state, status) {
    state.fetchPending = status;
  },

  FETCH_INFO(state, info) {
    state.fetchInfo = info;
  },
};

export const actions: ActionTree<UserState, UserState> = {
  async fetch({ commit, state }, context) {
    commit("FETCH_PENDING", true);
    try {
      const gender = context?.gender;
      const nat = context?.nat;

      const { data } = await HTTPClient.get<Users>("/", {
        params: {
          results: 50,
          seed: context?.seed,
          page: context?.page || 1,
          nat,
          gender,
        },
      });

      if (data?.results) {
        commit("SET_USERS", [...(state.users || []), ...data.results]);
      }

      if (data?.info) {
        commit("FETCH_INFO", {
          ...data.info,
          ...(gender && { gender }),
          ...(nat && { nat }),
        });
      }
    } catch {
      commit("FETCH_INFO", null);
    } finally {
      await sleep(400);
      commit("FETCH_PENDING", false);
    }
  },
};
