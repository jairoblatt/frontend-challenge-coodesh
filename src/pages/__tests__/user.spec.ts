import { mount, MountOptions, Wrapper, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import * as AlertModule from "@/store/modules/alert";
import * as UsersModule from "@/store/modules/users";
import User from "../user.vue";
import { MockUser } from "@/utils/tests";
import VueMeta from "vue-meta";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueMeta);

Object.entries({ Btn, Icon, Avatar, Spinner, ContentInfo, UrlClipBoard }).forEach(
  ([componentName, component]) => {
    localVue.component(componentName, component);
  }
);

type Instance = InstanceType<typeof User>;
let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

const $route = {
  params: {
    username: MockUser.login.username,
  },
};

const $router = {
  push: jest.fn(),
};

const goToUsersSpy = jest.spyOn((User as any).options.methods, "goToUsers");

describe("User.vue ", () => {
  beforeEach(() => {
    const store = new Vuex.Store({
      modules: {
        users: {
          namespaced: true,
          ...UsersModule,
          state: {
            users: [MockUser],
          },
        },
        // eslint-disable-next-line
        // @ts-ignore
        alert: {
          namespaced: true,
          ...AlertModule,
        },
      },
    });

    mountFunction = (options = {}) => {
      return mount(User, {
        localVue,
        store,
        mocks: {
          $route,
          $router,
        },
        ...options,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should mount the avatar component when", () => {
    const wrapper = mountFunction();

    const avatarComponent = wrapper.findComponent(Avatar);
    const imageElement = avatarComponent.find("img");

    expect(imageElement.attributes()).toEqual({
      src: MockUser.picture.large,
      alt: MockUser.name.first + " avatar",
    });
    expect(avatarComponent.exists()).toBe(true);
  });

  it('Should call "goToUsers" method and next call the push router method when button is clicked', () => {
    const wrapper = mountFunction();

    const btnComponent = wrapper.findComponent(Btn);
    btnComponent.trigger("click");

    expect(goToUsersSpy).toHaveBeenCalled();
    expect($router.push).toHaveBeenCalled();
  });

  it('Should render 6 "ContentInfo" component', () => {
    const wrapper = mountFunction();

    expect(wrapper.findAllComponents(ContentInfo)).toHaveLength(6);
  });

  it('Should render the "ContentInfo" id when user has id', () => {
    const wrapper = mountFunction();

    const contentInfoComponent = wrapper.find('[data-testid="idContentInfo"]');

    expect(contentInfoComponent.text()).toBe(`${MockUser.id.name}: ${MockUser.id.value}`);
  });

  it('Should set alert message when "userFilteredByUsername" length is 0', async () => {
    const wrapper = mountFunction();

    wrapper.vm.$store.commit("users/SET_USERS", []);

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.$store.getters["alert/get"]).toBeTruthy();
    expect(goToUsersSpy).toHaveBeenCalled();
    expect($router.push).toHaveBeenCalled();
  });

  it("Should set username in the page meta", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm.$meta().refresh().metaInfo.title).toBe(
      `${MockUser.name.first} ${MockUser.name.last}`
    );
  });
});
