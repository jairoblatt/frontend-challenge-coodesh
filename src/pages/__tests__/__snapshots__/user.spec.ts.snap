// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User.vue  Should mount the component 1`] = `
"<div class=\\"modal\\">
  <div class=\\"modal__user\\">
    <picture class=\\"avatar user__avatar\\" style=\\"height: 12rem; width: 12rem; border-width: 4px;\\">
      <source srcset=\\"https://randomuser.me/api/portraits/thumb/women/6.jpg\\" media=\\"(max-width: 639px)\\">
      <source srcset=\\"https://randomuser.me/api/portraits/med/women/6.jpg\\" media=\\"(min-width: 640px) and (max-width: 1023px)\\"> <img src=\\"https://randomuser.me/api/portraits/women/6.jpg\\" alt=\\"Liva avatar\\"></picture>
    <div class=\\"modal__header\\"><button class=\\"btn\\" data-testid=\\"modal-close\\">
        <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi  mdi-close\\" style=\\"font-size: 21px;\\"></i>
      </button></div>
    <div class=\\"user__content\\">
      <div class=\\"content__name\\">
        <h1>Liva Christiansen</h1>
      </div>
      <div class=\\"content__andress\\">
        <h2><i class=\\"mdi  mdi-map-marker-outline \\" style=\\"font-size: 14px;\\"></i>
          Stoevring, Syddanmark
          <span>(Denmark)</span></h2>
        <h3><i class=\\"mdi mdi-home-variant-outline\\" style=\\"font-size: 14px;\\"></i>
          Hirsev√¶nget, 5831
        </h3>
      </div>
      <hr>
      <div class=\\"content__infos\\">
        <div class=\\"content__info\\"><span><i class=\\"mdi mdi-email-outline\\" style=\\"font-size: 19px;\\"></i></span> <small>
            liva.christiansen@example.com
          </small></div>
        <div class=\\"content__info\\"><span><i class=\\"mdi mdi-gender-male-female\\" style=\\"font-size: 19px;\\"></i></span> <small>
            female
          </small></div>
        <div class=\\"content__info\\"><span><i class=\\"mdi mdi-candle\\" style=\\"font-size: 19px;\\"></i></span> <small>
            51
          </small></div>
        <div class=\\"content__info\\"><span><i class=\\"mdi mdi-cellphone\\" style=\\"font-size: 19px;\\"></i></span> <small>
            41815273
          </small></div>
        <div class=\\"content__info\\"><span><i class=\\"mdi mdi-phone\\" style=\\"font-size: 19px;\\"></i></span> <small>
            66653454
          </small></div>
        <div class=\\"content__info\\" data-testid=\\"idContentInfo\\"><span><i class=\\"mdi mdi-card-account-details\\" style=\\"font-size: 19px;\\"></i></span> <small>
            CPR: 270570-1815
          </small></div>
      </div>
      <hr>
      <div class=\\"content__url\\"><span>Copy user url to clipboard:</span>
        <div class=\\"user__url\\"><input type=\\"text\\" readonly=\\"readonly\\"> <button class=\\"btn\\" data-testid=\\"clipboard\\">
            <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-content-copy\\" style=\\"font-size: 14px;\\"></i>
          </button></div>
      </div>
    </div>
  </div>
</div>"
`;
